### PR TITLE
Adds the ability to notifylisteners for values that don't change. Additionally adds a pipeResult function to pipe the results from one command to another.

### PR DESCRIPTION
**Pull request by Abhilash-Chandran**
This PR aims to fix the issues pointed out in issue #8 and also adds the pipeResults function to enable passing results of one command to another without explicitly listening. 

Added additional tests to cover the new logic added. All existing tests passed. Two of the existing tests were modified to accommodate the new changes. This is especially due to the changes in `_ListenerCountingValueNotifier` which also notifies when there is no change and `results` is of this type.

Other existing ValueListenables such as `isExecuting`, `canExecute` and `thrownExceptions` haven't been modified.

- [ ] Documentation needs to be updated once the solution is agreed and finalized.

### CODE REVIEW

**Comment by escamoteur** [comment]
Hi, sorry I totally missed this PR

**Comment by Abhilash-Chandran** [comment]
> Hi, sorry I totally missed this PR

No worries😃. Take your time.

File: lib/flutter_command.dart [hunk 1]
```diff
@@ -103,7 +103,8 @@ class CommandError<TParam> {
 /// where [TParam] is the type of data that is passed when calling [execute] and
 /// [TResult] denotes the return type of the handler function. To signal that
 /// a handler doesn't take a parameter or returns no value use the type `void`
-abstract class Command<TParam, TResult> extends ValueNotifier<TResult> {
+abstract class Command<TParam, TResult>
+    extends _ValueEqualityNotifier<TResult> {
```
Comment (escamoteur) [inline lib/flutter_command.dart]: I added a `CustomValueNotifier` that does the same thing and its already used for thrownExceptions. Please use thaat one

File: lib/flutter_command.dart [hunk 2]
```diff
@@ -421,6 +447,42 @@ abstract class Command<TParam, TResult> extends ValueNotifier<TResult> {
...
+  /// [pipedCommand] accepts parameters of the same type as [TPipedCommandParam].
+  ///
+  /// As a fall back case, [pipedCommandParam] will be used if the [value]
+  /// returned by the command doesn't match the parameter type of the
+  /// [pipedCommand]. This allows connecting non compatible commands if needed.
+  ///
+  Command pipeResult<TPipedCommandParam>(
+    Command pipedCommand, {
+    @required bool pipedCommandTakesParam,
+    TPipedCommandParam pipedCommandParam,
```
Comment (escamoteur) [inline lib/flutter_command.dart]: Better calling is `pipedDefaultParam`

File: lib/flutter_command.dart [hunk 3]
```diff
@@ -421,6 +447,42 @@ abstract class Command<TParam, TResult> extends ValueNotifier<TResult> {
...
+    Duration debounceDuration = const Duration(),
+  }) {
+    assert(pipedCommand != null, 'pipedCommand cannot be null');
+    assert(pipedCommandTakesParam != null,
+        'pipedCommandTakesParam cannot be null');
+    debounce(debounceDuration).listen((TResult resultValue, _) {
```
Comment (escamoteur) [inline lib/flutter_command.dart]: where does this `debounce` come from?
Oh wait, you used the debounce extension. please write `this.debounce` I really took a moment.

**Review by escamoteur — CHANGES_REQUESTED** [review]
Looks good in general. Unfortunately I did the null migration in parallel. 
Could you please update it with things I commented and see that it merges into the null safety branch?
thanks a lot
Thomas

**Comment by Abhilash-Chandran** [comment]
Sure I already looked into your comments. Will do the required changes. I am currently busy with my thesis mid term review. Perhaps I will work on this by mid of next week and update the PR. 😄

File: lib/flutter_command.dart (hunk 3 above)
Comment (Abhilash-Chandran) [reply lib/flutter_command.dart:474]: I was getting a dart lint hint for this saying unnecessary this.

```
Don't access members with `this` unless avoiding shadowing. dart(unnecessary_this)
```

Comment (Abhilash-Chandran) [reply lib/flutter_command.dart]: I was getting a dart lint hint saying unnecessary this. The debounce comes from functional_listener I believe.

```
Don't access members with `this` unless avoiding shadowing. dart(unnecessary_this)
```

File: lib/flutter_command.dart (hunk 2 above)
Comment (Abhilash-Chandran) [reply lib/flutter_command.dart]: Will do. Makes sense . 👍

File: lib/flutter_command.dart (hunk 1 above)
Comment (Abhilash-Chandran) [reply lib/flutter_command.dart]: Sorry I didn't get your hint here. 😃 
The one used for `thrownExceptions` is to count the listeners right. But this functionality is not required for the Command I assume. Hence I went with the idea of having a separate `_ValueEqualityNotifier`. I will look into this further in new NullSafe version. 👍

Comment (escamoteur) [reply lib/flutter_command.dart]: Nope, the counting one is the `_commandResult` see
grafik

File: lib/flutter_command.dart (hunk 3 above)
Comment (escamoteur) [reply lib/flutter_command.dart]: yeah, please disable this lint in analysis_option. sometimes especially in cases like this it's confusing not to use `this.` if you use an extension method. It took me several minutes to finally get it

File: lib/flutter_command.dart (hunk 1 above)
Comment (Abhilash-Chandran) [reply lib/flutter_command.dart]: Ok this is in the updated version. I just pulled the repo to verify where `CustomValueNotifier` comes from. So its from `functional_listener`. I think it shouldn't be difficult to use that. Will work on this next week. 👍

**Comment by escamoteur** [comment]
Hi, are you still looking after this PR?

**Comment by Abhilash-Chandran** [comment]
> Hi, are you still looking after this PR?

Hi Thomas, I am sorry for the late reply. I am nearing my master thesis deadline and I have been busy for last one month with the writing and experiments. I should be free by 14th June.  After that I will take this up. I know this shouldn't take more than a day to fix and write tests. But I am truly busy. Otherwise I would have finished it long ago. :)

**Comment by escamoteur** [comment]
NP, just wanted to check up what's the status

**Comment by Abhilash-Chandran** [comment]
Hi Thomas, I updated the commit with the recommended changes. Could you verify them once. Existing tests are all passing. Also, 3 existing tests were already failing so didn't try to handle it in this PR. I think we can handle it as a separate issue.

**Comment by escamoteur** [comment]
@Abhilash-Chandran I'm deeply sorry not to have cared earlier about this but the second half of 2021 wasn't a good time for my mental health.
It looks great , but I think we need a possibility to removed a piped commands, so we have to keep track of the commands that are added to the pipe.
We also need to update the docs accordingly. Especially the signaling even if the value stays the same is a breaking change.
Would you still be willing to help with this package

**Comment by Abhilash-Chandran** [comment]
Hi Thomas,

I get you. I am glad you are back at this. I was a bit occupied with studies and new job and all.

Regarding the removal of pipes it totally makes sense. I will see how to do that over the weekend and update the PR.

I will also add the documentation for this logic.

**Comment by Abhilash-Chandran** [comment]
Hi Thomas,

I was thinking about tracking the pipedCommands. My initial idea was to hold the `ListenableSubscription` which will be returned while setting up the listener and then cancel it when the user wants to remove the pipedCommand.

But this raised some questions.

1. Do we allow multiple pipe commands to be added in parallel, to the same command.
```dart
Command a;
Command b;
Command c;

a.pipeResult(b);
a.pipeResult(c);
```
    - If yes then how do we enable removing specific pipedCommands.
2. Or do we allow only one command to be piped to another command and overwrite.

```dart
a.pipeResult(b);

// following line cancels existing pipe and adds c as the newly piped command.
a.pipeResult(c);
```

Since we call this a pipe for me the second one makes more sense. Then we only need to track one `ListenableSubscription`.

**Comment by escamoteur** [comment]
Let's start with one command and see if people start to ask for more
Am 6. Feb. 2022, 17:38 +0100 schrieb Abhilash Chandran ***@***.***>:

**Comment by Abhilash-Chandran** [comment]
Hi Thomas, 

I think we should handle the `pipeResult` logic in a separate issue or PR. I just tested it and it's very easy to create a deadlock scenario without a clear way to avoid piping two commands to each other accidentally. So I removed the logic for  now including the tests. I have updated the docs.

For e.g.

```dart
test("Create Infinite callback", () async {
  final Command rootCommand = Command.createSync<String, String>((s) {
    print('inside root command ===> $s');
    return s;
  }, "initialValue");

  // This command takes list of string and returns a string concatenated by
  // as under-score '_'.
  final Command pipedCommand = Command.createSync<String, String>((input) {
    final _val = '${input}_';
    print('inside piped command ===> $_val');
    return _val;
  }, "initial value");

  final Collector pipedCommandResults = Collector<String>();

  // Note here the type of the result from root command is not the same
  // as parameter of pipedCommand.
  rootCommand.pipeResult<String>(
    pipedCommand,
    pipedCommandTakesParam: true,
  );

  pipedCommand.listen((result, _) {
    pipedCommandResults(result);
  });
  // // Create a inifite call loop.
  pipedCommand.pipeResult<String>(
    rootCommand,
    pipedCommandTakesParam: true,
  );
  rootCommand("Root Value");
  await Future.delayed(const Duration(milliseconds: 50));

});
```

File: lib/flutter_command.dart [hunk 4]
```diff
@@ -841,3 +873,22 @@ class _ListenerCountingValueNotifier<T> extends ValueNotifier<T> {
     listenerCount = 0;
```
Comment (escamoteur) [inline lib/flutter_command.dart:873]: I'm pretty sure this is wrong if it should keep track of the number of listeners. I could add the listener count in my CustomNotifier

**Review by escamoteur — COMMENTED** [review]
besides my comment above it looks fine. Tell me if I should add the listener count

**Comment by escamoteur** [comment]
On the pipeResult. I think we could add a function that lets you check if a command is already is piped. That you'd be used to avoid deadlocks.
But I agree we should put it in a separate PR

**Comment by escamoteur** [comment]
Sorry, that I lost track of this here again. Where are we in the current process?

**Pull request merged**